src/decode/lzma.rs

```diff
@@ -264,76 +302,212 @@ where
-                    let idx: usize;
-                    if !rangecoder.decode_bit(&mut self.is_rep_g1[self.state])? {
-                        idx = 1;
-                    } else if !rangecoder.decode_bit(&mut self.is_rep_g2[self.state])? {
-                        idx = 2;
-                    } else {
-                        idx = 3;
-                    }
+                    idx = 3;
+                }
+                if update {
                     // Update LRU
                     let dist = self.rep[idx];
                     for i in (0..idx).rev() {
                         self.rep[i + 1] = self.rep[i];
                     }
                     self.rep[0] = dist
                 }
+            }
+
+            len = self.rep_len_decoder.decode(rangecoder, pos_state, update)?;
 
-                len = self.rep_len_decoder.decode(rangecoder, pos_state)?;
+            if update {
                 // update state (rep)
                 self.state = if self.state < 7 { 8 } else { 11 };
-            // New distance
-            } else {
+            }
+        // New distance
+        } else {
+            if update {
                 // Update LRU
                 self.rep[3] = self.rep[2];
                 self.rep[2] = self.rep[1];
                 self.rep[1] = self.rep[0];
-                len = self.len_decoder.decode(rangecoder, pos_state)?;
+            }
 
+            len = self.len_decoder.decode(rangecoder, pos_state, update)?;
+
+            if update {
                 // update state (match)
                 self.state = if self.state < 7 { 7 } else { 10 };
-                self.rep[0] = self.decode_distance(rangecoder, len)?;
+            }
 
+            let rep_0 = self.decode_distance(rangecoder, len, update)?;
+
+            if update {
+                self.rep[0] = rep_0;
                 if self.rep[0] == 0xFFFF_FFFF {
                     if rangecoder.is_finished_ok()? {
-                        break;
+                        return Ok(false);
                     }
                     return Err(error::Error::LZMAError(String::from(
                         "Found end-of-stream marker but more bytes are available",
                     )));
                 }
             }
+        }
 
+        if update {
             len += 2;
 
             let dist = self.rep[0] + 1;
             self.output.append_lz(len, dist)?;
         }
 
+        Ok(true)
+    }
+
+    fn process_next<'a, R: io::BufRead>(
+        &mut self,
+        rangecoder: &mut rangecoder::RangeDecoder<'a, R>,
+    ) -> error::Result<bool> {
+        self.process_next_inner(rangecoder, true)
+    }
+
+    /// Try to process the next iteration of the loop.
+    ///
+    /// This will check to see if there is enough data to consume and advance the
+    /// decompressor. Needed in streaming mode to avoid corrupting the state while
+    /// processing incomplete chunks of data.
+    #[allow(clippy::if_same_then_else)]
```

Comment:
Why is this clippy lint allowed?

Reply:
I was able to remove it without any issues.